### PR TITLE
feat: Distributed checkpointing

### LINKED ISSUE #109: Improve HF checkpointing time with torch dist checkpointing

### LINKED ISSUE #110: Support AutoModel.from_pretrained from reinforcer checkpoints
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

Provide a code snippet on how new APIs/changes would be used by others.

**Describe alternatives you've considered**

**Additional context**

Add any other context or screenshots about the feature request here

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

Closes #109
Closes #110 

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/reinforcer/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/reinforcer/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/reinforcer/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 1]
```diff
@@ -786,80 +788,23 @@ def move_to_cpu(self, model):
...
-            scheduler_state_dict = None
-        with FullyShardedDataParallel.state_dict_type(
-            self.model,
-            state_dict_type=StateDictType.FULL_STATE_DICT,
-            state_dict_config=state_dict_config,
-        ):
-            # Load model weights
-            self.model.load_state_dict(state_dict if state_dict else None)
+        state_dict = {
+            "experiment": ExpState(self.model, self.optimizer, self.scheduler)
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/hf_policy.py]: is this nesting required for dist ckpting?

File: nemo_reinforcer/algorithms/sft.py [hunk 2]
```diff
@@ -175,10 +175,7 @@ def setup(
     policy = HfPolicy(
         cluster=cluster,
         config=policy_config,
-        weights_path=Path(last_checkpoint_path) / "policy.pt"
-        if last_checkpoint_path
-        else None,
-        optimizer_path=Path(last_checkpoint_path) / "policy_optimizer.pt"
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/sft.py:181]: since we now bundle everything together, does this mean we cannot selectively load just the policy?

File: nemo_reinforcer/models/policy/hf_policy.py (hunk 1 above)
Comment (ashors1) [reply nemo_reinforcer/models/policy/hf_policy.py]: I wasn't able to get checkpoint restore to work without the nesting and the dist checkpoint tutorial does it this way, but I'm not sure whether it's actually required. I will look into it and verify

File: nemo_reinforcer/algorithms/sft.py (hunk 2 above)
Comment (ashors1) [reply nemo_reinforcer/algorithms/sft.py:181]: Currently, we cannot, but we could easily add a flag that determines whether to load the optimizer states from the checkpoint

Comment (terrykong) [reply nemo_reinforcer/algorithms/sft.py:181]: It would be great if we could still make sure we could selectively load stuff since (1) it makes publishing easier (can just nuke `rm policy_optimizer.pt`) and (2) ensure fast and mem efficient load if we don't need the optimizer

(1) is more for UX, but do we have a path forward for (2)?

Comment (ashors1) [reply nemo_reinforcer/algorithms/sft.py:181]: This is definitely possible. I think we'd need to do something similar to [what's done in mcore](https://github.com/NVIDIA/Megatron-LM/blob/6ba97dd37150a6bfba03d31808674211cf2a4d0d/megatron/core/dist_checkpointing/strategies/filesystem_async.py#L129-L147) where we provide a `separation_hint` and write everything whose prefix starts with that `separation_hint` to a separate file. I'll need to look into this a bit more to understand the complexity involved

Comment (ashors1) [reply nemo_reinforcer/algorithms/sft.py:181]: As a simpler solution, I'm just saving the optimizer and model weights as two separate steps. This allows us to specify different paths for the model weights and optimizer states. Let me know what you think of this

Comment (terrykong) [reply nemo_reinforcer/algorithms/sft.py:181]: how does the ckpt save/load time look when doing it this way?

Comment (ashors1) [reply nemo_reinforcer/algorithms/sft.py:181]: Checkpoint save time seems marginally slower when using this approach vs saving model weights and optimizer states together (e.g. on 8 H100s, saving everything together takes roughly 18.3s while saving separately takes around 19.3s, but that's based on a relatively small amount of data). Checkpoint load time is about the same

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 3]
```diff
@@ -787,79 +793,41 @@ def move_to_cpu(self, model):
     def save_checkpoint(
         self,
         weights_path: str,
-        optimizer_path: Optional[str] = None,
-        offload_to_cpu: bool = True,
+        optimizer_path: str,
```
Comment (SahilJain314) [inline nemo_reinforcer/models/policy/hf_policy.py]: any reason to remove the optionality here?

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 4]
```diff
@@ -787,79 +793,41 @@ def move_to_cpu(self, model):
...
-            optimizer_state_dict = None
-            scheduler_state_dict = None
-        with FullyShardedDataParallel.state_dict_type(
-            self.model,
-            state_dict_type=StateDictType.FULL_STATE_DICT,
-            state_dict_config=state_dict_config,
-        ):
-            # Load model weights
-            self.model.load_state_dict(state_dict if state_dict else None)
+    def load_checkpoint(self, weights_path: str, optimizer_path: str):
```
Comment (SahilJain314) [inline nemo_reinforcer/models/policy/hf_policy.py]:
```suggestion
    def load_checkpoint(self, weights_path: str, optimizer_path: Optional[str] = None):
```

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 5]
```diff
@@ -880,7 +848,6 @@ def __init__(
     ):
         if weights_path:
             weights_path = os.path.abspath(weights_path)
-        if optimizer_path:
```
Comment (SahilJain314) [inline nemo_reinforcer/models/policy/hf_policy.py]: likewise here on optionality

File: nemo_reinforcer/utils/checkpoint.py [hunk 6]
```diff
@@ -26,6 +26,73 @@
...
+from torch.distributed.checkpoint.state_dict import (
+    get_model_state_dict,
+    set_model_state_dict,
+    get_optimizer_state_dict,
+    set_optimizer_state_dict,
+)
+
+
+## modified from pytorch tutorial https://pytorch.org/tutorials/recipes/distributed_checkpoint_recipe.html
+class ModelState(Stateful):
```
Comment (SahilJain314) [inline nemo_reinforcer/utils/checkpoint.py]: Do you think these are general across backends or should they go in a specific FSDP1/HF checkpoint utility file?

File: nemo_reinforcer/models/policy/hf_policy.py (hunk 3 above)
Comment (ashors1) [reply nemo_reinforcer/models/policy/hf_policy.py]: Nope, just got lost with all the changes. I added it back

File: nemo_reinforcer/utils/checkpoint.py (hunk 6 above)
Comment (ashors1) [reply nemo_reinforcer/utils/checkpoint.py]: These are specific to pure HF/pytorch. I moved them to a hf-specific checkpoint file

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 7]
```diff
@@ -788,78 +794,41 @@ def save_checkpoint(
...
-            state_dict_config=state_dict_config,
-        ):
-            # Save model state dict
-            model_state_dict = self.model.state_dict()
-            optim_state_dict = FullyShardedDataParallel.optim_state_dict(
-                self.model, self.optimizer
-            )
-            scheduler_state_dict = self.scheduler.state_dict()
+        model_state_dict = {"model": ModelState(self.model)}
+        dcp.save(model_state_dict, checkpoint_id=weights_path)
```
Comment (parthchadha) [inline nemo_reinforcer/models/policy/hf_policy.py]: Can we move the save/load of distributed checkpoint to `hf_checkpoint.py`. It will be a self contained file containing all checkpointing logic.

File: examples/convert_dcp_to_hf.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+from nemo_reinforcer.utils.config import load_config
+
+
+def parse_args():
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(
+        description="Convert Torch DCP checkpoint to HF checkpoint"
+    )
+    parser.add_argument(
+        "--config", type=str, default=None, help="Path to YAML config file"
```
Comment (terrykong) [inline examples/convert_dcp_to_hf.py]: should this be the one in the checkpoint? if so, could you remind users?

File: examples/convert_dcp_to_hf.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    dcp_ckpt = args.dcp_ckpt_path
+    hf_ckpt = args.hf_ckpt_path
+
+    # Extract individual configs for easier access
+    policy_config = config["policy"]
+    cluster_config = config["cluster"]
+
+    cluster = RayVirtualCluster(
+        name="sft_cluster",
```
Comment (terrykong) [inline examples/convert_dcp_to_hf.py]: should this say `sft`?

File: examples/convert_dcp_to_hf.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        config = OmegaConf.merge(config, overrides)
+
+    dcp_ckpt = args.dcp_ckpt_path
+    hf_ckpt = args.hf_ckpt_path
+
+    # Extract individual configs for easier access
+    policy_config = config["policy"]
+    cluster_config = config["cluster"]
+
+    cluster = RayVirtualCluster(
```
Comment (terrykong) [inline examples/convert_dcp_to_hf.py:63]: can we do this without the cluster/gpus? btw, if we have to use the cluster, then it's always good to call `init_ray()` first

File: nemo_reinforcer/utils/hf_checkpoint.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+        optimizer: Optional optimizer to save
+        scheduler: Optional scheduler to save
+        optimizer_path: Path to save optimizer state (required if optimizer provided)
+        save_torch_dist: Whether to save in PyTorch distributed format
+        save_hf: Whether to save in HuggingFace format
+    """
+    if save_hf:
+        # Create a new path by appending "-hf" to the weights path
+        hf_weights_path = f"{Path(weights_path)}-hf"
+        model.save_pretrained(hf_weights_path)
```
Comment (terrykong) [inline nemo_reinforcer/utils/hf_checkpoint.py]: are all ranks doing this?

File: examples/convert_dcp_to_hf.py (hunk 10 above)
Comment (ashors1) [reply examples/convert_dcp_to_hf.py:63]: With the current implementation, I think it's simplest to use the cluster. I want to take advantage of `HFPolicy`'s checkpoint save and load functionality, but since the model is moved to GPU during init, we need multiple GPUs to avoid OOM with larger models. This can definitely be avoided with some refactoring, but I was hoping to keep the code changes required for this converter to a minimum

File: nemo_reinforcer/utils/hf_checkpoint.py (hunk 11 above)
Comment (ashors1) [reply nemo_reinforcer/utils/hf_checkpoint.py]: Yeah.. I don't think this should be required, but when I try to save only on rank 0, I get a hang. I'll see if I can figure out what the problem is

Comment (ashors1) [reply nemo_reinforcer/utils/hf_checkpoint.py]: The hang happens when trying to get the state dict: https://github.com/huggingface/transformers/blob/5290f6a62dd4892b07ff290087591d8a625a9f1f/src/transformers/modeling_utils.py#L2879. It looks like we need all ranks to participate in getting the state dict when the model is sharded using FSDP. @SahilJain314 maybe you would know something about this from your work on FDSP1 checkpointing?

File: nemo_reinforcer/models/policy/hf_policy.py (hunk 1 above)
Comment (ashors1) [reply nemo_reinforcer/models/policy/hf_policy.py]: I think we need the values in our state dict to be Stateful objects in order for DCP to call state_dict and load_state_dict when needed

File: examples/convert_dcp_to_hf.py (hunk 8 above)
Comment (ashors1) [reply examples/convert_dcp_to_hf.py]: No, this should be the yaml config file used during training. I updated the description to clarify this. Let me know if it's still unclear.

Comment (terrykong) [reply examples/convert_dcp_to_hf.py]: So right now it says:
```
Path to YAML config file used during model training
```
would it be better to say
```
Path to YAML config file in checkpoint directory
```
If the user overrides something on the CLI, won't the one in the checkpoint be the most up to date?

File: examples/convert_dcp_to_hf.py (hunk 10 above)
Comment (terrykong) [reply examples/convert_dcp_to_hf.py:63]: can you create an issue for a non-gpu way of converting? I am okay with this for now

If users train a large model like 70b we should offer a way to convert the model without needing the same level of compute as the training.

File: nemo_reinforcer/utils/hf_checkpoint.py (hunk 11 above)
Comment (terrykong) [reply nemo_reinforcer/utils/hf_checkpoint.py]: could you jsut pass in the state dict to avoid materializing and saving on all ranks?

https://github.com/huggingface/transformers/blob/5290f6a62dd4892b07ff290087591d8a625a9f1f/src/transformers/modeling_utils.py#L2681

File: docs/design_docs/checkpointing.md
Comment (terrykong) [inline docs/design_docs/checkpointing.md:1]: could you add this to `index.md` so it shows up in our docs?

File: examples/configs/grpo_math_1B.yaml [hunk 12]
```diff
@@ -25,6 +25,7 @@ checkpointing:
 
 policy:
   model_name: "meta-llama/Llama-3.2-1B-Instruct"
+  tokenizer_name: ${policy.model_name} ## specify if you'd like to use a tokenizer different from the model's default
```
Comment (terrykong) [inline examples/configs/grpo_math_1B.yaml:28]: could you add this new key to the config to help with typing?

https://github.com/NVIDIA/reinforcer/blob/main/nemo_reinforcer/models/policy/__init__.py#L21

File: examples/convert_dcp_to_hf.py (hunk 8 above)
Comment (ashors1) [reply examples/convert_dcp_to_hf.py]: Okay yeah, I was expecting users to pass the exact yaml config used in training rather than the json config saved to the checkpoint, but I can update the code to use the json config instead

Comment (terrykong) [reply examples/convert_dcp_to_hf.py]: i also created an issue for persisting the git sha into the checkpoint metadata:

https://github.com/NVIDIA/reinforcer/issues/130

Related to this PR since I can tell users will be frustrated trying to find the working converter as it evolves

File: examples/convert_dcp_to_hf.py (hunk 10 above)
Comment (terrykong) [reply examples/convert_dcp_to_hf.py:63]: https://github.com/NVIDIA/reinforcer/issues/142

Comment (SahilJain314) [reply examples/convert_dcp_to_hf.py:63]: It would be good to offer, but it makes sense to have the gpu path be default. Iirc, conversion is absurdly slow on cpu

File: nemo_reinforcer/utils/hf_checkpoint.py
Comment (terrykong) [inline nemo_reinforcer/utils/hf_checkpoint.py:1]: should we rename this? maybe native_checkpoint.py or something? there's not really anything in here about transformers

File: tests/functional/sft.sh [hunk 13]
```diff
@@ -26,10 +26,15 @@ python -u $PROJECT_ROOT/examples/run_sft.py \
...
     logger.wandb_enabled=false \
-    checkpointing.enabled=false \
+    checkpointing.enabled=true \
+    checkpointing.save_every_n_steps=10 \
+    checkpointing.checkpoint_dir=/tmp/sft_checkpoints \
     $@ \
     2>&1 | tee $RUN_LOG
 
+## clean up checkpoint directory
+rm -r /tmp/sft_checkpoints
```
Comment (terrykong) [inline tests/functional/sft.sh]: it's better to do `trap` here b/c if the job fails you don't clean up

File: nemo_reinforcer/algorithms/grpo.py [hunk 14]
```diff
@@ -613,6 +613,12 @@ def grpo_train(
                 and (step + 1) % master_config["checkpointing"]["save_period"] == 0
             ):  # +1 because step is 0-indexed
                 policy.prepare_for_training()
+
+                is_last_checkpoint = (
+                    master_config["grpo"]["max_num_steps"] - (step + 1)
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/grpo.py]: The max steps in the config is an upper bound. If the dataset is small relative to it, training will exit early ( we only do 1 epoch). (I.e dataset has 100 steps worth of data vs max steps at 10000)

File: nemo_reinforcer/algorithms/sft.py [hunk 15]
```diff
@@ -399,19 +397,25 @@ def sft_train(
                 master_config["checkpointing"]["enabled"]
                 and (step + 1) % master_config["checkpointing"]["save_period"] == 0
             ):  # +1 because step is 0-indexed
-                sft_save_state["step"] = step
+                is_last_checkpoint = (
+                    master_config["sft"]["max_num_steps"] - (step + 1)
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/sft.py]: Less familiar with sft. Do you do multiple epochs here?

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 16]
```diff
@@ -811,78 +817,29 @@ def save_checkpoint(
...
         weights_path: str,
         optimizer_path: Optional[str] = None,
-        offload_to_cpu: bool = True,
+        save_torch_dist: bool = True,
+        save_hf: bool = False,
     ):
-        # Config to save full state dict on rank 0, offloaded to CPU
-        state_dict_config = FullStateDictConfig(
-            offload_to_cpu=offload_to_cpu, rank0_only=True
+        """Save a checkpoint of the model."""
```
Comment (SahilJain314) [inline nemo_reinforcer/models/policy/hf_policy.py]: Since the logic of the structure of the checkpoint is getting pushed into an external function, would you mind describing the layout here?

File: nemo_reinforcer/algorithms/sft.py (hunk 15 above)
Comment (ashors1) [reply nemo_reinforcer/algorithms/sft.py]: Good point. SFT currently only does a single epoch but we should support multiple epochs. I made an issue for this which I'll address in a separate PR: https://github.com/NVIDIA/reinforcer/issues/143.

For now, I updated this check to take the min of `master_config["sft"]["max_num_steps"]` and `len(dataloader)`

File: nemo_reinforcer/algorithms/grpo.py (hunk 14 above)
Comment (ashors1) [reply nemo_reinforcer/algorithms/grpo.py]: Updated to take the min of `master_config["grpo"]["max_num_steps"]` and `len(datalaoder)`
